### PR TITLE
[TLS Lib Clean-up] Add host name check For tls library

### PR DESCRIPTION
**Pull request by ZhenLian**
This PR is intended to add host name check in current tls credential stack, since we already have the constants set in https://github.com/grpc/grpc/blob/2b200d2313e739fa1841edf3ea86ab8532862505/include/grpc/grpc_security_constants.h#L121-L133 but the checking logic is not complete, if requiring host name checking.
The final goal is to migrate the underneath ssl credential stack to tls credential stack.

### CODE REVIEW

**Comment by ZhenLian** [comment]
@yihuazhang  
Could you please take a look at this PR? Thank you so much for the initial review!

File: src/core/lib/security/security_connector/tls/tls_security_connector.cc [hunk 1]
```diff
@@ -112,6 +112,18 @@ grpc_status_code TlsFetchKeyMaterials(
   return status;
 }
 
+grpc_error* TlsCheckPeer(const char* peer_name, const tsi_peer* peer) {
```
Comment (yihuazhang) [inline src/core/lib/security/security_connector/tls/tls_security_connector.cc]: s/TlsCheckPeer/TlsCheckHostName/g

File: src/core/lib/security/security_connector/tls/tls_security_connector.cc [hunk 2]
```diff
@@ -180,6 +192,17 @@ void TlsChannelSecurityConnector::check_peer(
       grpc_ssl_peer_to_auth_context(&peer, GRPC_TLS_TRANSPORT_SECURITY_TYPE);
   const TlsCredentials* creds =
       static_cast<const TlsCredentials*>(channel_creds());
+  if (creds->options().server_verification_option() ==
```
Comment (yihuazhang) [inline src/core/lib/security/security_connector/tls/tls_security_connector.cc]: Please also set the default value for `server_verification_option_` in `grpc_tls_credentials_options.h`

Comment (yihuazhang) [reply src/core/lib/security/security_connector/tls/tls_security_connector.cc]: And make sure hostname verification check is enabled in e2e tests.

**Comment by yihuazhang** [comment]
Thanks Zhen for the PR. It looks good to me overall. Some minor comments are left.

**Comment by ZhenLian** [comment]
First Round of tests, 4 failed/47 passed tests:

- Bazel RBE Windows Opt C/C++:  
  - Failing on `//test/core/end2end:h2_ssl_test@retry_non_retriable_status_before_recv_trailing_metadata_started`
  - possible issue: https://github.com/grpc/grpc/issues/20922
- Basic Tests C/C++ MacOS: 
  - Failing on XdsTest/DropTest.Update/FakeResolver, where GetParam() = 2-byte object <00-00>
  - Issue: https://github.com/grpc/grpc/issues/21230
- Basic Tests ObjC MacOS
  - Failing on CallAPIv2Tests testTimeout
  - Issue: https://github.com/grpc/grpc/issues/20686
- Bazel C/C++ Opt MacOS
  - Failing on `//test/core/channel:channelz_test`
  - Issue: https://github.com/grpc/grpc/issues/20324

I am going to rerun this for one more time and see how the tests behave.

**Comment by ZhenLian** [comment]
Second round of tests, 3 failed/ 48 passed tests:
- Bazel C/C++ Opt MacOS
  - Failing on `//test/core/end2end:h2_census_test@retry_exceeds_buffer_size_in_initial_batch`
  - Possible issue: https://github.com/grpc/grpc/issues/13609
- Bazel RBE Opt C/C++ 
  - Failing on `//test/cpp/end2end:xds_end2end_test@poller=epoll1`
  - Possible issue: https://github.com/grpc/grpc/issues/20695
- Bazel RBE UBSAN C/C++
  - Failing on `//test/cpp/end2end:xds_end2end_test@poller=epoll1`
  - Issue: https://github.com/grpc/grpc/issues/20695

**Comment by ZhenLian** [comment]
Looks like none of the failed tests are relevant to this PR. I will merge it later.

**Pull request merged**